### PR TITLE
protobuild: add support for descriptor aggregation

### PR DESCRIPTION
**Pull request by stevvooe**
This change introduces the ability to aggregate descriptors into
arbitrary collections under project root. Descriptors outline the
services and messages that make up a GRPC service. Aggregating all of
the GRPC service definitions into a single file provides output that can
be used to ensure that changes don't impact API stability.

Signed-off-by: Stephen J Day <stephen.day@docker.com>

@crosbymichael @dmcgowan @mlaventure @estesp PTAL

### CODE REVIEW

File: descriptors.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+type descriptorSet struct {
+	merged      descriptor.FileDescriptorSet
+	seen        map[string]struct{}
+	ignoreFiles map[string]struct{}
+}
+
+func newDescriptorSet(ignoreFiles ...string) *descriptorSet {
+	ifm := make(map[string]struct{}, len(ignoreFiles))
+	for _, ignore := range ignoreFiles {
+		fmt.Println(ignore, "ignore")
```
Comment (mlaventure) [inline descriptors.go]: left over debug?

File: main.go [hunk 2]
```diff
@@ -111,6 +149,47 @@ func main() {
...
+			}
+
+			for path, set := range descriptorSets {
+				if strings.HasPrefix(importDirPath, path) {
+					set.add(desc.File...)
+				}
+			}
+			descriptors = append(descriptors, desc)
+
+			// clean up descriptors file
```
Comment (mlaventure) [inline main.go]: Move the close and the remove in a defer after `dfp` is created I'd say

File: Protobuild.toml [hunk 3]
```diff
@@ -43,3 +43,64 @@ plugins = ["grpc"]
...
+# generated descriptor file will include both the "foo" and "bar" subpackages.
+prefix = "github.com/stevvooe/protobuild/examples"
+
+# Target defines the output location of the merged descriptor file.
+target = "examples/next.pb.txt"
+
+# Ignore files by their protobuf import path.
+#
+# This operates directly on the FileDescriptorProto.Name field, so it will
+# filter out imported protobufs. This allows ony to remove descriptors from the
```
Comment (dmcgowan) [inline Protobuild.toml]: type: `ony`

File: main.go (hunk 2 above)
Comment (stevvooe) [inline main.go]: Doing so might leave too many open file descriptors.

**Pull request merged**